ForwardedRequestCustomizer behavior should not be applied to requests without forwarding headers

joakime opened the pull request:
If the request has no Forwarding headers, the customization that ForwardingRequestCustomizer typically does should be skipped.

Closes #5443 

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5443 - Forwarding Headers are optional

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java:
```diff
@@ -461,81 +461,88 @@ public void customize(Connector connector, HttpConfiguration config, Request req
+            if (forwarded._proto != null)
+            {
+                proto = forwarded._proto;
+                request.setScheme(proto);
```
why setting here and not below?

joakime replied:
The `forwarded._proto` tracking value is set from 3 different scenarios.

1. `X-Forwarded-Proto: <value>` header.
2. `X-Proxied-Https: on` sets it to `https`.
3. `Forwarding: proto=<value>` header.

So this one that you highlighted is correct.

The below action is to handle the not-quite-forwarding-headers oddball cases.

https://github.com/eclipse/jetty.project/blob/0b646ee6b78c5e4e1dff6ddbf641b839360d8a87/jetty-server/src/test/java/org/eclipse/jetty/server/ForwardedRequestCustomizerTest.java#L731-L743

Which has the `ForwardingRequestCustomizer.setSslIsSecure(true)` which should trigger both the proto/scheme to `https` and the `request.setSecure(true)` if the `Proxy-Ssl-Id` or `Proxy-auth-cert` headers are present.

Let me see if I can simplify this a bit.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java:
```diff
@@ -461,81 +461,88 @@ public void customize(Connector connector, HttpConfiguration config, Request req
+            // Set secure status
+            if (forwarded.isSecure() ||
+                proto.equalsIgnoreCase(config.getSecureScheme()) ||
+                port == getSecurePort(config))
```
I'm don't think that just because it arrives on the secure port is sufficient to make it secure.  Some ports might do both secure and insecure requests.

gregw reviewed (changes requested):
The proto and secure logic is spread out too much and too convoluted.
I think it should just be in one spot and should be something like:
```java
    if (forwarded._proto != null)
    {
      request.setScheme(forwarded._proto);
      if (forwarded._proto.equalsIgnoreCase(config.getSecureScheme())
        request.setSecure(true);
    }
    if (forwarded.isSecure())
    {
        request.setSecure(true);
        if (forwarded._proto == null)
          request.setScheme(config.getSecureScheme());
    }
```

that might not be exactly right... but you get the idea.  Put it all in one spot and try to separate concerns as much as possible.

gregw commented:
Isn't this for #5437 not #5443 ???

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java:
```diff
@@ -461,81 +461,88 @@ public void customize(Connector connector, HttpConfiguration config, Request req
-
-        // Is secure status configured from headers?
-        if (forwarded.isSecure())
+        if (match)
```
I don't think this match logic should be required.   There are many headers and some are orthogonal to others, so the logic below must be robust and can't be gated by a match, which may be for an unrelated header.

In the OPs case, proto must not be changed unless there is an explicit header that changes it - either giving a proto or forcing secure.

Perhaps this match is an optimisation to avoid several ifs below, but I don't think it is necessary as this customizer is typically only deployed when all request will have such a header.

joakime replied:
The match logic here (in jetty-9.4.x) was taken from jetty-10.0.x.

https://github.com/eclipse/jetty.project/blame/jetty-10.0.x/jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java#L464-L482

The idea behind it is that if there's no Forwarding headers (even those not-quite-forwarding headers `X-Proxied-Https`, `Proxy-auth-cert`, and `Proxy-ssl-id` that ForwardingRequestCustomizer also pays attention to) then there should be no action taken by ForwardingRequestCustomizer, as the incoming request is not based on forwarding.

joakime commented:
> Isn't this for #5437 not #5443 ???

Issue #5437 is about this kind of request ...

```
GET https://localhost:12345 HTTP/1.1
Host: localhost
```

A request with no forwarding headers, a fully qualified URL in the request-line, with https.
Which altered the scheme/proto in the wrong way.

Which is a (modified) test case in this PR ...

https://github.com/eclipse/jetty.project/blob/0b646ee6b78c5e4e1dff6ddbf641b839360d8a87/jetty-server/src/test/java/org/eclipse/jetty/server/ForwardedRequestCustomizerTest.java#L178-L186

This testcase is just to prevent regression in the future.  

But the fix was actually handled in PR #5419 

The fix to not act if there are no forwarding headers, which is what this PR covers, also addresses this above example, not changing the request.metadata if there's no forwarding headers. (so in essence issue #5437 was fixed in two different ways).

Issue #5443 is about an NPE when there are no forwarding headers and the Host is empty (example in issue is given in HTTP/1.0)

Which is two test cases in this PR ...

https://github.com/eclipse/jetty.project/blob/0b646ee6b78c5e4e1dff6ddbf641b839360d8a87/jetty-server/src/test/java/org/eclipse/jetty/server/ForwardedRequestCustomizerTest.java#L138-L157

joakime committed:
Issue #5443 - Forwarding Headers are optional

+ Simplify isSecure handling in customize.
+ Simplify handling of `Proxy-Ssl-Id` header.
+ Simplify handling of `Proxy-auth-cert` header.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5443 - Forwarding Headers are optional

+ Improve / document implied secure scheme behaviors
  for both `Proxy-Ssl-Id` or `Proxy-auth-cert`

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5443 - Forwarding Headers are optional

+ Additional tests for HTTP/1.0
+ Overly complex negative test cases for
   `X-Forwarded-Proto: http` and
   `X-Proxied-Https: off`
+ Failure testcase for `X-Proxied-Https: foo`

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5443 - Forwarding Headers are optional

Additional NPE safety checks.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5443 - Forwarding Headers are optional

The `X-Proxied-Https: off` case should have an implied port
not a hardcoded port.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5443 - Forwarding Headers are optional

Cleanup handling of forwarded.authority

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

bentmann commented:
Thank you!
